pypeit/reduce.py

```diff
@@ -1081,3 +1138,394 @@ def local_skysub_extract(self, global_sky, sobjs,
+        fluxarr = np.append(fluxarr, scale_dict['scale'][:, 2][wmx] * sclfct_mx)
+
+        # Create the final interpolating polynomial, and apply it to the wavelength image
+        refspl = interp1d(wavearr, fluxarr, kind='cubic', bounds_error=False, fill_value="extrapolate")
+        scale_model = refspl(self.waveimg)
+
+        # Now return
+        return self.caliBrate.flatimages.flat_model/scale_model
+
+    def get_platescale(self, dummy):
+        """
+        Return the platescale for IFU.
+        The input argument is ignored
+
+        Args:
+            dummy (:class:`pypeit.specobj.SpecObj`):
+                ignored
+
+        Returns:
+            float:
+
+        """
+        plate_scale = self.sciImg.detector.platescale
+        return plate_scale
+
+    def resample_cube(self):
+        pass
+
+    def load_skyregions(self):
+        skymask_init = None
+        if self.par['reduce']['skysub']['load_mask']:
+            # Check if a master Sky Regions file exists for this science frame
+            file_base = os.path.basename(self.sciImg.files[0])
+            prefix = os.path.splitext(file_base)
+            if prefix[1] == ".gz":
+                sciName = os.path.splitext(prefix[0])[0]
+            else:
+                sciName = prefix[0]
+
+            # Setup the master frame name
+            master_dir = self.caliBrate.master_dir
+            master_key = self.caliBrate.fitstbl.master_key(0, det=self.det) + "_" + sciName
+
+            regfile = masterframe.construct_file_name(buildimage.SkyRegions,
+                                                      master_key=master_key,
+                                                      master_dir=master_dir)
+            # Check if a file exists
+            if os.path.exists(regfile):
+                msgs.info("Loading SkyRegions file for: {0:s} --".format(sciName) + msgs.newline() + regfile)
+                skymask_init = fits.getdata(regfile).astype(np.bool)
+            else:
+                msgs.warn("SkyRegions file not found:" + msgs.newline() + regfile)
+        elif self.par['reduce']['skysub']['user_regions'] != '':
+            msgs.info("Generating skysub mask based on the user defined regions: {0:s}".format(
+                self.par['reduce']['skysub']['user_regions']))
+            # This doesn't need to be a user parameter
+            resolution = int(10.0*np.max(self.slits_right-self.slits_left))
+            # Get the regions
+            status, regions = skysub.read_userregions(self.par['reduce']['skysub']['user_regions'],
+                                                      resolution=resolution)
+            # Generate image
+            skymask_init = skysub.generate_mask(self.pypeline, regions, self.slits, self.slits_left, self.slits_right,
+                                                resolution=resolution)
+        return skymask_init
+
+    def run(self, basename=None, ra=None, dec=None, obstime=None,
```

Comment:
Do you have to overload `run`?  I'm guessing yes,
but kindly explain why.   Will be a bit painful to
maintain the two of them..

Reply:
Agreed... unfortunately, I think this will be needed. Perhaps we will need to be more modular in the original `run()` function, so we can apply certain features but not all. Basically, for IFU, you don't want to find/extract  objects. You just need to extract the entire slit/FOV.

For now, I've removed all of the IFU code from the `reduce.py` code so that it can be the focus of the next PR. This PR is really about bringing all the calibrations up to speed with the rest of the code.